modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java

```diff
@@ -461,31 +466,33 @@ private void consumeSegment(Path segment) {
                 }
             }
 
-            builder.from(initState);
-
-            initState = null;
+            builder.from(initState.pointer());
         }
 
-        try (WALIterator it = factory.iterator(builder)) {
-            boolean interrupted = Thread.interrupted();
+        try (DataEntryIterator iter = new DataEntryIterator(factory.iterator(builder))) {
+            if (initState != null) {
+                iter.init(initState);
 
-            while (it.hasNext() && !interrupted) {
-                Iterator<DataRecord> iter = F.iterator(it.iterator(), t -> (DataRecord)t.get2(), true);
+                initState = null;
+            }
+
+            boolean interrupted = Thread.interrupted();
 
+            while (iter.hasNext() && !interrupted) {
                 boolean commit = consumer.onRecords(iter);
 
                 if (commit) {
-                    assert it.lastRead().isPresent();
+                    CdcState curState = iter.state();
```

Comment:
Can we move `state` method to the `WalRecordsConsumer` to keep iteration logic inside it?

Reply:
The state is relative to the wal iterator.